geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/GeoMessage.scala

```diff
@@ -0,0 +1,184 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.locationtech.geomesa.kafka
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+
+import com.typesafe.scalalogging.slf4j.Logging
+import org.joda.time.Instant
+import org.locationtech.geomesa.features.SerializationOption.SerializationOptions
+import org.locationtech.geomesa.features.kryo.KryoFeatureSerializer
+import org.locationtech.geomesa.features.{SimpleFeatureDeserializer, SimpleFeatureSerializer}
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+sealed trait GeoMessage {
+
+  def timestamp: Instant
+}
+
+object GeoMessage {
+
+  /** @return a new [[Clear]] message with the current time */
+  def clear(): Clear = Clear(Instant.now)
+
+  /** @return a new [[Clear]] message with the given feature ``id`` and current time */
+  def delete(id: String): Delete = Delete(Instant.now, id)
+
+  /** @return a new [[CreateOrUpdate]] message with the given ``sf`` and current time */
+  def createOrUpdate(sf: SimpleFeature): CreateOrUpdate = CreateOrUpdate(Instant.now, sf)
+}
+
+/** Create a new simple feature or update an existing [[SimpleFeature]]
+  *
+  * @param feature the [[SimpleFeature]]
+  */
+case class CreateOrUpdate(override val timestamp: Instant, feature: SimpleFeature) extends GeoMessage {
+
+  val id = feature.getID
+}
+
+/** Delete an existing [[SimpleFeature]]
+  *
+  * @param id the id of the simple feature
+ */
+case class Delete(override val timestamp: Instant, id: String) extends GeoMessage
+
+/** Delete all [[SimpleFeature]]s
+  */
+case class Clear(override val timestamp: Instant) extends GeoMessage
+
+
+/** Encodes [[GeoMessage]]s.  [[Clear]] and [[Delete]] messages are handled directly.  See class
+  * [[GeoMessageEncoder]] for encoding [[CreateOrUpdate]] messages.
+  *
+  * The following encoding is used:
+  *
+  * Key: version (1 byte) type (1 byte) timstamp (8 bytes)
+  *
+  * The current version is 1
+  * The type is 'C' for create or update, 'D' for delete and 'X' for clear
+  *
+  * Value
+  *   CreateOrUpdate: serialized simple feature
+  *   Delete: simple feature ID as UTF-8 bytes
+  *   Clear: empty
+  *
+  */
+object GeoMessageEncoder {
+
+  val version: Byte = 1
+  val createOrUpdateType: Byte = 'C'
+  val deleteType: Byte = 'D'
+  val clearType: Byte = 'X'
+
+  private val EMPTY = Array.empty[Byte]
+
+  def encodeKey(msg: GeoMessage): Array[Byte] = {
+
+    val msgType: Byte = msg match {
+      case c: CreateOrUpdate => createOrUpdateType
+      case d: Delete => deleteType
+      case x: Clear => clearType
+      case u => throw new IllegalArgumentException(s"Invalid message: '$u'")
```

Comment:
It already is a sealed trait.  The last case is mainly to handle null although it should not be null...